client/allocrunner/taskrunner/plugin_supervisor_hook.go

```diff
@@ -0,0 +1,323 @@
+	h.runner.hookResources.setMounts(mounts)
+
+	resp.Done = true
+	return nil
+}
+
+// Poststart is called after the task has started. Poststart is not
+// called if the allocation is terminal.
+//
+// The context is cancelled if the task is killed.
+func (h *csiPluginSupervisorHook) Poststart(_ context.Context, _ *interfaces.TaskPoststartRequest, _ *interfaces.TaskPoststartResponse) error {
+	// If we're already running the supervisor routine, then we don't need to try
+	// and restart it here as it only terminates on `Stop` hooks.
+	h.runningLock.Lock()
+	if h.running == true {
+		h.runningLock.Unlock()
+		return nil
+	}
+	h.runningLock.Unlock()
+
+	go h.ensureSupervisorLoop(h.shutdownCtx)
+	return nil
+}
+
+// ensureSupervisorLoop should be called in a goroutine. It will terminate when
+// the passed in context is terminated.
+//
+// The supervisor works by:
+// - Initially waiting for the plugin to become available. This loop is expensive
+//   and may do things like create new gRPC Clients on every iteration.
+// - After receiving an initial healthy status, it will inform the plugin catalog
+//   of the plugin, registering it with the plugins fingerprinted capabilities.
+// - We then perform a more lightweight check, simply probing the plugin on a less
+//   frequent interval to ensure it is still alive, emitting task events when this
+//   status changes.
+//
+// Deeper fingerprinting of the plugin is implemented by the csimanager.
+func (h *csiPluginSupervisorHook) ensureSupervisorLoop(ctx context.Context) {
+	h.runningLock.Lock()
+	if h.running == true {
+		h.runningLock.Unlock()
+		return
+	}
+	h.running = true
+	h.runningLock.Unlock()
+
+	defer func() {
+		h.runningLock.Lock()
+		h.running = false
+		h.runningLock.Unlock()
+	}()
+
+	socketPath := filepath.Join(h.mountPoint, structs.CSISocketName)
+	t := time.NewTimer(0)
+
+	// Step 1: Wait for the plugin to initially become available.
+WAITFORREADY:
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			pluginHealthy, err := h.supervisorLoopOnce(ctx, socketPath)
+			if err != nil || !pluginHealthy {
+				h.logger.Info("CSI Plugin not ready", "error", err)
+				t.Reset(5 * time.Second)
+				continue
+			}
+
+			// Mark the plugin as healthy in a task event
+			h.previousHealthState = pluginHealthy
+			event := structs.NewTaskEvent(structs.TaskPluginHealthy)
+			event.SetMessage(fmt.Sprintf("plugin: %s", h.task.CSIPluginConfig.ID))
+			h.eventEmitter.EmitEvent(event)
+
+			break WAITFORREADY
+		}
+	}
+
+	// Step 2: Register the plugin with the catalog.
+	deregisterPluginFn, err := h.registerPlugin(socketPath)
+	if err != nil {
+		h.logger.Error("CSI Plugin registration failed", "error", err)
+		event := structs.NewTaskEvent(structs.TaskPluginUnhealthy)
+		event.SetMessage(fmt.Sprintf("failed to register plugin: %s, reason: %v", h.task.CSIPluginConfig.ID, err))
+		h.eventEmitter.EmitEvent(event)
```

Comment:
We can't fail a task here, because the task would already have started. Or we have to block in Poststart, but that makes a bunch of allocrunner behaviour sad (iirc it stopped us from doing shutdown and stuff correctly when I tried, and I don't wanna refactor it now)

Reply:
Hm, seems like we should file an issue for ourselves to fix that at some point. Won't this leave the plugin in a "looks to be running fine but can't actually service requests" state? Seems like a worst case scenario, but maybe I'm missing something.